standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/properties/Digester.java

```diff
@@ -0,0 +1,440 @@
+        return this;
+    }
+
+    /**
+     * Updates the digest with a bytes array.
+     * @param bytes the bytes
+     * @return this digester
+     */
+    public Digester digest(byte[] bytes) {
+        if (bytes != null) {
+            md.update(bytes);
+        }
+        return this;
+    }
+
+    /**
+     * Updates the digest with an integer.
+     * @param i4 the int
+     * @return this digester
+     */
+    public Digester digest(int i4) {
+        md.update((byte) (i4 & 0xff));
+        md.update((byte) (i4 >> 8));
+        md.update((byte) (i4 >> 16));
+        md.update((byte) (i4 >> 24));
+        return this;
+    }
+
+    /**
+     * Updates the digest with a long.
+     * @param l8 the long
+     * @return this digester
+     */
+    public Digester digest(long l8) {
+        digest(md, l8);
+        return this;
+    }
+
+    /**
+     * Updates the digest with a double.
+     * @param f8 the double
+     * @return this digester
+     */
+    public Digester digest(double f8) {
+        digest(md, Double.doubleToRawLongBits(f8));
+        return this;
+    }
+
+    /**
+     * Updates the digest with a date.
+     * @param date the date
+     * @return this digester
+     */
+    public Digester digest(Date date) {
+        if (date != null) {
+            digest(md, date.getTime());
+        }
+        return this;
+    }
+
+    /**
+     * Updates the digest with a string.
+     * @param str the string
+     * @return this digester
+     */
+    public Digester digest(String str) {
+        if (str != null) {
+            final int sz = str.length();
+            for(int i = 0; i < sz; ++i) {
+                digest(str.charAt(i));
+            }
+        }
+        return this;
+    }
+
+    /**
+     * Updates the digest with a uuid.
+     * @param uid the uuid
+     * @return this digester
+     */
+    public Digester digest(UUID uid) {
+        digest(md, uid);
+        return this;
+    }
+
+    /**
+     * Updates the digest with a uuid.
+     * @param uri the uri
+     * @return this digester
+     */
+    public Digester digest(URI uri) {
+        digest(md, uri.toString());
+        return this;
+    }
+
+    /**
+     * Updates the digest with an object that describes how it digests.
+     * @param digestible the object
+     * @return this digester
+     */
+    public Digester digest(Digestible digestible) {
+        if (digestible != null) {
+            digestible.digest(this);
+        }
+        return this;
+    }
+
+    /**
+     * Updates the digest with a stream.
+     * @param input the stream
+     * @return this digester
+     */
+    public Digester digest(InputStream input) {
+        if (input != null) {
+            digest(md, input);
+        }
+        return this;
+    }
+    /**
+     * Updates the digest with any (serializable) object.
+     * @param obj the object
+     * @return this digester
+     */
+    public Digester digest(Object obj) {
+        if (obj instanceof Digestible) {
```

Comment:
Do you think this can be better implemented in the switch case?

Reply:
Might try and switch on the class name in some concrete cases but would still need a few 'ifs'. Besides, Digestible classes and/or direct calls make this generic method rarely needed.

Reply:
I'm reluctant to try and optimize a method that has not been measured as being a problem. Furthermore, I would expect the 10  'if instanceof'  versus a hash look up to be in the same performance range.